pants-plugins/api_spec/rules.py

```diff
@@ -0,0 +1,259 @@
+from pants.core.target_types import FileSourceField, ResourceSourceField
+from pants.core.util_rules.source_files import SourceFiles, SourceFilesRequest
+from pants.engine.addresses import Address
+from pants.engine.fs import (
+    CreateDigest,
+    Digest,
+    FileContent,
+    MergeDigests,
+    Snapshot,
+)
+from pants.engine.process import FallibleProcessResult, ProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import (
+    FieldSet,
+    SourcesField,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.util.logging import LogLevel
+
+from api_spec.target_types import APISpecSourceField
+
+
+# these constants are also used in the tests
+CMD_SOURCE_ROOT = "st2common"
+CMD_DIR = "st2common/st2common/cmd"
+CMD_MODULE = "st2common.cmd"
+GENERATE_CMD = "generate_api_spec"
+VALIDATE_CMD = "validate_api_spec"
+
+
+@dataclass(frozen=True)
+class APISpecFieldSet(FieldSet):
+    required_fields = (APISpecSourceField,)
+
+    source: APISpecSourceField
+
+
+class GenerateAPISpecViaFmtTargetsRequest(FmtTargetsRequest):
+    field_set_type = APISpecFieldSet
+    name = GENERATE_CMD
+
+
+class ValidateAPISpecRequest(LintTargetsRequest):
+    field_set_type = APISpecFieldSet
+    name = VALIDATE_CMD
+
+
+@rule(
+    desc="Update openapi.yaml with st2-generate-api-spec",
+    level=LogLevel.DEBUG,
+)
+async def generate_api_spec_via_fmt(
+    request: GenerateAPISpecViaFmtTargetsRequest,
+) -> FmtResult:
+    # There will only be one target+field_set, but we iterate
+    # to satisfy how fmt expects that there could be more than one.
+    # If there is more than one, they will all get the same contents.
+
+    # Find all the dependencies of our target
+    transitive_targets = await Get(
+        TransitiveTargets,
+        TransitiveTargetsRequest(
+            [field_set.address for field_set in request.field_sets]
+        ),
+    )
+
+    dependency_files_get = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[
+                tgt.get(SourcesField) for tgt in transitive_targets.dependencies
+            ],
+            for_sources_types=(FileSourceField, ResourceSourceField),
+        ),
+    )
+
+    source_files_get = Get(
+        SourceFiles,
+        SourceFilesRequest(field_set.source for field_set in request.field_sets),
+    )
+
+    # actually generate it with an external script.
+    # Generation cannot be inlined here because it needs to import the st2 code.
+    pex_get = Get(
+        VenvPex,
+        PexFromTargetsRequest(
+            [
+                Address(
+                    CMD_DIR,
+                    target_name="cmd",
+                    relative_file_path=f"{GENERATE_CMD}.py",
+                ),
+            ],
+            output_filename=f"{GENERATE_CMD}.pex",
+            internal_only=True,
+            main=EntryPoint.parse(f"{CMD_MODULE}.{GENERATE_CMD}:main"),
```

Comment:
We are using `st2common.cmd.generate_api_spec:main` directly instead of using the `st2-generate-api-spec` script because pants can't run python with a `-` in the name (because pants relies on importing from the target script, and python cannot import from files with a `-` in the name).

Eventually I'll get around to plumbing a change in pants to make that possible (it already works with the underlying pex, but pants hasn't learned about that arg yet). In the meantime, we just use the entrypoint itself instead of the script.

Reply:
Having a command with `-` is fairly common, so I don't think that's a problem.

Pex (used by pants) already had everything it needed to read files without importing them, but that didn't have a user facing API. So, when I asked about it, John (pex maintainer) quickly added an `--executable` option, so now pex supports it. But, pants hasn't learned about that pex option yet, so it doesn't use it.

I started adding the bits here, but haven't had time to dive farther than this: https://github.com/pantsbuild/pants/compare/main...cognifloyd:pants:pex-exe-arg